packages/commonwealth/server/util/requirementsModule/validateGroupMembership.ts

```diff
@@ -21,7 +23,114 @@ export default async function validateGroupMembership(
   requirements: Requirement[],
   tbc?: TokenBalanceCache
 ): Promise<validateGroupMembershipResponse> {
-  return {
+  const response: validateGroupMembershipResponse = {
     isValid: true,
+    messages: [],
   };
+  const checks = requirements.map(async (requirement) => {
+    let checkResult: { result: boolean; message: string };
+    switch (requirement.rule) {
+      case 'threshold': {
+        checkResult = await _thresholdCheck(userAddress, requirement.data, tbc);
+        break;
+      }
+      case 'allow': {
+        checkResult = await _allowlistCheck(
+          userAddress,
+          requirement.data as AllowlistData
+        );
+        break;
+      }
+      default:
+        checkResult = {
+          result: false,
+          message: 'Invalid Requirment',
+        };
+        break;
+    }
+    if (!checkResult.result) {
+      response.isValid = false;
+      response.messages.push({
+        requirement,
+        message: checkResult.message,
+      });
+    }
+  });
+  await Promise.all(checks);
+  return response;
+}
+
+async function _thresholdCheck(
+  userAddress: string,
+  thresholdData: ThresholdData,
+  tbc: TokenBalanceCache
+): Promise<{ result: boolean; message: string }> {
+  try {
+    let chainNetwork: ChainNetwork;
+    let contractAddress: string;
+    let chainId: string;
+    switch (thresholdData.source.source_type) {
+      case 'erc20': {
+        chainNetwork = ChainNetwork.ERC20;
+        contractAddress = thresholdData.source.contract_address;
+        chainId = thresholdData.source.evm_chain_id.toString();
+        break;
+      }
+      case 'erc721': {
+        chainNetwork = ChainNetwork.ERC721;
+        contractAddress = thresholdData.source.contract_address;
+        chainId = thresholdData.source.evm_chain_id.toString();
+        break;
+      }
+      case 'eth_native': {
+        chainNetwork = ChainNetwork.Ethereum;
+        chainId = thresholdData.source.evm_chain_id.toString();
+        break;
+      }
+      case 'cosmos_native': {
+        chainNetwork = ChainNetwork.Osmosis;
```

Comment:
Can we leave a comment explaining if/why this ChainNetwork setting is OK for any Cosmos native token?

Reply:
commented